examples/06-multiple-connections/server.c

```diff
@@ -0,0 +1,455 @@
+				pdata.len, sizeof(rpma_mr_descriptor));
+		(void) rpma_conn_disconnect(clnt->conn);
+		return;
+	}
+	desc = (rpma_mr_descriptor *)pdata.ptr;
+
+	/* prepare a remote memory region */
+	struct rpma_mr_remote *src_mr;
+	int ret = rpma_mr_remote_from_descriptor(desc, &src_mr);
+	if (ret) {
+		print_error_ex("rpma_mr_remote_from_descriptor", ret);
+		(void) rpma_conn_disconnect(clnt->conn);
+		return;
+	}
+
+	/* read client's name from the remote memory region */
+	ret = rpma_read(clnt->conn, svr->dst_mr, clnt->offset,
+			src_mr, 0, DWORD, RPMA_F_COMPLETION_ALWAYS, NULL);
+	if (ret) {
+		print_error_ex("rpma_read", ret);
+		(void) rpma_mr_remote_delete(&src_mr);
+		(void) rpma_conn_disconnect(clnt->conn);
+	}
+
+	/* delete the remote memory region's object */
+	ret = rpma_mr_remote_delete(&src_mr);
+	if (ret)
+		print_error_ex("rpma_read", ret);
+}
+
+/*
+ * conn_event_closed -- callback on connection is closed
+ *
+ * Delete:
+ * - all client's file descriptors from the server's epoll
+ * - the connection
+ */
+void
+conn_event_closed(struct client_res *clnt)
+{
+	struct server_res *svr = clnt->svr;
+
+	(void) epoll_delete(svr->epoll, &clnt->ev_conn_cmpl);
+	(void) epoll_delete(svr->epoll, &clnt->ev_conn_event);
+
+	/* delete the connection */
+	int ret = rpma_conn_delete(&clnt->conn);
+	if (ret)
+		print_error_ex("rpma_conn_delete", ret);
+
+	/* decrease the number of active connections */
+	--svr->client_num;
+}
+
+/*
+ * conn_event -- callback on connection's next event
+ */
+void
+conn_event(struct custom_event *ce)
+{
+	struct client_res *clnt = (struct client_res *)ce->arg;
+
+	/* get next connection's event */
+	enum rpma_conn_event event;
+	int ret = rpma_conn_next_event(clnt->conn, &event);
+	if (ret) {
+		if (ret == RPMA_E_NO_NEXT)
+			return;
+
+		print_error_ex("rpma_conn_next_event", ret);
+		(void) rpma_conn_disconnect(clnt->conn);
+		return;
+	}
+
+	/* proceed to the callback specific to the gotten event */
+	switch (event) {
+	case RPMA_CONN_ESTABLISHED:
+		conn_event_established(clnt);
+		break;
+	case RPMA_CONN_CLOSED:
+	default:
+		conn_event_closed(clnt);
+	}
+}
+
+/*
+ * ep_incoming_func -- callback on endpoint's next incoming connection
+ *
+ * Get the connection request. If there is not free slots reject it. Otherwise,
+ * accept the incoming connection, get the event and completion file
+ * descriptors, set O_NONBLOCK flag for both of them and add events to
+ * the epoll.
+ * If error will occur at any of the required steps the client is disconnected.
+ */
+void
+ep_incoming_func(struct custom_event *ce)
+{
+	struct server_res *svr = (struct server_res *)ce->arg;
+
+	/* receive an incoming connection request */
+	struct rpma_conn_req *req = NULL;
+	int ret = rpma_ep_next_conn_req(svr->ep, &req);
+	if (ret) {
+		if (ret == RPMA_E_NO_NEXT)
+			return;
+
+		print_error_ex("rpma_ep_next_conn_req", ret);
+		return;
+	}
+
+	/* check if at least one free slot is available */
+	if (svr->client_num == CLIENT_MAX) {
```

Comment:
This is not necessary. If the loop fails to find a NULL client then you know there are no free slots.

In general, avoid duplication of state if possible. Any duplication requires you to keep things in-sync, and is just a source of mistakes.